jvm/src/main/scala/cats/jvm/EvalAsync.scala

```diff
@@ -0,0 +1,41 @@
+package cats
+package jvm
+
+import data.Reader
+import java.util.concurrent.{Callable, CountDownLatch, ExecutorService}
+
+object EvalAsync {
+  def apply[A](cb: (A => Unit) => Unit): Eval[A] = Eval.always {
+    val cdl = new CountDownLatch(1)
+    var result: Option[A] = None
+    cb((a: A) => {result = Some(a); cdl.countDown})
+    cdl.await
```

Comment:
This is not asynchronous and it's not equivalent to a `Task`, because this is cheating :-)

I think Cats should strive to expose the same API for both the JVM and Scala.js. For one because it would allow people to write portable code without worrying about which types and functions are available where. But also because it keeps you honest.

When you do a `countDownLatch.await`, in essence blocking a thread for a result, the JVM pretends that the processing you're doing is synchronous, with a result available immediately. But it's a trick that can backfire. Ever heard of EJB (Enterprise Java Beans)? Ever heard of Corba? THE reason for why these sucked so badly was because their original design was a game of pretend. As in, lets pretend that where the processing happens for these synchronous methods don't matter. But of course, making local and remote method calls look the same ended up being a problem, because in the real world you can have significant latency, bandwidth limitations, security issues, dropped packages, timing issues and all sorts of other networking problems.

But back to the point, blocking threads is dangerous and in order to block safely, you have to know the configuration of the underlying thread-pool. Lets say that our thread-pool is limited to a single thread. Lets say that our `countDownLatch.await` will block that single thread. At that point there will be no threads left to do the actual processing. A one-thread pool is simplifying the problem of course, but this can happen with *any fixed sized thread-pool*. 

This is why we have a [BlockContext](https://github.com/alexandru/scala-best-practices/blob/master/sections/4-concurrency-parallelism.md#44-must-use-scalas-blockcontext-on-blocking-io) in Scala, being able to *tell the thread-pool* that this and that operations are potentially blocking so you might want to add more threads or whatever. Such a strategy is acceptable for blocking I/O. But this is very inefficient, as it defeats the purpose of using a limited thread-pool, so when doing a lot of synchronous I/O (like for example JDBC stuff), we might as well use a `CachedThreadPool` with unlimited threads.

Which leads us to the reason for why we want to use limited thread pools (e.g. with a number of threads directly proportional to the number of CPU cores). It's because preemptive multi-threading of more threads than CPU-cores is done by a technique called [time slicing](https://en.wikipedia.org/wiki/Super-threading) and when a CPU core switches between threads/processes it has to do a [context switch](https://en.wikipedia.org/wiki/Context_switch) and context switches are freaking expensive. This is why it's considered a best practice to do async I/O and to do CPU-bound processing on a limited thread-pool and if you have to do synchronous I/O, then people instantiate a second unlimited thread-pool meant only for synchronous I/O.

In other words, on top of the JVM, we have 3 options:

1. thread-block a limited thread-pool and suffer from non-deterministic *deadlocks*
2. thread-block an unlimited thread-pool and suffer the performance penalties
3. do not block any threads, ever

So you see, from my point of view, this isn't even an issue related to Javascript :-)